query-engine/core/src/query_graph_builder/extractors/filters/scalar.rs

```diff
@@ -126,12 +283,27 @@ where
     F: Fn(Filter) -> Filter,
 {
     let inner_object: ParsedInputMap = input.try_into()?;
-    let mut results = vec![];
+    let filters: Vec<Filter> = parse(inner_object, field, reverse)?;
 
-    for (k, v) in inner_object {
-        let filters = parse(&k, field, v, reverse)?;
-        results.extend(filters);
-    }
+    Ok(filters.into_iter().map(|filter| func(filter)).collect())
+}
+
+fn parse_json_path(input: ParsedInputValue) -> QueryGraphBuilderResult<JsonFilterPath> {
+    let path: PrismaValue = input.try_into()?;
 
-    Ok(results.into_iter().map(|filter| func(filter)).collect())
+    match path {
+        PrismaValue::String(str) => Ok(JsonFilterPath::String(str)),
+        PrismaValue::List(list) => {
+            let keys = list
+                .into_iter()
+                .map(|key| {
+                    key.into_string()
+                        .expect("Json filtering array path elements must all be of type string")
```

Comment:
You have the right return type to throw a real error here. You don't need to if the validation stage already guarantees the correctness, in which case this unwrap is fine. I know that JSON is dicey in the code, which is why I'd ask you to take a look if you can actually guarantee correctness here. If you can't, you need to throw an actual error.